siliconcompiler/dependency.py

```diff
@@ -0,0 +1,178 @@
+    """
+
+    # Initially try retrieving dependency from schema
+    dependency = {}
+    dependency['path'] = chip.get('dependency', package, 'path')
+    dependency['ref'] = chip.get('dependency', package, 'ref')
+    if not dependency['path']:
+        # If not in the schema retrieve dependency from the python package and store in schema
+        dependency = importlib.import_module(package).dependency
+        for k, v in dependency.items():
+            chip.set('dependency', package, k, v)
+
+    if not dependency.get('path'):
+        chip.logger.error('A valid path needs to be specified')
+    url = urlparse(dependency['path'])
+
+    # check network drive for dependency data
+    if dependency['path'].startswith('file://') or os.path.exists(dependency['path']):
+        path = dependency['path'].replace('file://', '')
+        if not quiet:
+            chip.logger.info(f'Found {dependency} data at {path}')
+        return path
+
+    # location of the python package
+    cache_path = os.path.join(Path.home(), '.sc', 'cache')
+    if not os.path.exists(cache_path):
+        os.makedirs(cache_path, exist_ok=True)
+    project_id = f'{package}-{dependency.get("ref")}'
+    if url.scheme not in ['git', 'git+https', 'https', 'git+ssh', 'ssh'] or not project_id:
+        chip.error(f'Could not find dependency data in package {package}')
+    data_path = os.path.join(cache_path, project_id)
+
+    # Wait a maximum of 10 minutes for other git processes to finish
+    lock_file = Path(data_path+'.lock')
+    wait_on_lock(chip, data_path, lock_file, max_seconds=600)
+
+    # check cached dependency data
+    if os.path.exists(data_path):
+        if not quiet:
+            chip.logger.info(f'Found cached {package} data at {data_path}')
+        if url.scheme in ['git', 'git+https', 'ssh', 'git+ssh']:
+            try:
+                lock_file.touch()
+                repo = Repo(data_path)
+                if not quiet and (repo.untracked_files or repo.index.diff("HEAD")):
+                    chip.logger.warning('The repo of the cached data is dirty.')
+                return data_path
+            except GitCommandError:
+                chip.logger.warning('Deleting corrupted cache data.')
+                shutil.rmtree(path)
+            finally:
+                lock_file.unlink(missing_ok=True)
+        else:
+            return data_path
+
+    # download dependency data
+    if url.scheme in ['git', 'git+https', 'ssh', 'git+ssh']:
+        clone_synchronized(chip, package, dependency, data_path, lock_file)
+    elif url.scheme == 'https':
+        extract_from_url(chip, package, dependency, data_path)
+    if os.path.exists(data_path):
+        chip.logger.info(f'Saved {package} data to {data_path}')
+        return data_path
+    chip.logger.error(f'Extracting {package} data to {data_path} failed')
+    # Exit clean and early as missing dependencies would definitely cause further issues
+    exit(1)
```

Comment:
replace with `raise SiliconCompilerException`

Reply:
Sure I can do that.
My original intention with this was to show the user that there was an issue with their dependency.
The exception would make it more seem like there was an issue within SiliconCompiler.